plugins/radare2/radare2_main.ml

```diff
@@ -0,0 +1,90 @@
+open Bap_core_theory
+open Core_kernel
+open Bap_future.Std
+open Bap.Std
+include Self()
+
+open KB.Syntax
+
+let agent =
+  KB.Agent.register ~package:"bap.std" "radare2-symbolizer"
+    ~desc:"extracts symbols radare2"
+
+let provide_roots funcs =
+  let promise_property slot =
+    KB.promise slot @@ fun label ->
+    KB.collect Theory.Label.addr label >>| function
+    | None -> None
+    | Some addr ->
+      let addr = Bitvec.to_bigint addr in
+      Option.some_if (Hashtbl.mem funcs addr) true in
+  promise_property Theory.Label.is_valid;
+  promise_property Theory.Label.is_subroutine
+
+exception Radare2_failed
+
+let provide_radare2 file = 
+  let funcs = Hashtbl.create (module struct
+      type t = Z.t
+      let compare = Z.compare and hash = Z.hash
+      let sexp_of_t x = Sexp.Atom (Z.to_string x)
+    end) in
+    let accept name addr = Hashtbl.set funcs addr name in
+    let extract name json = let open Yojson in
+      match json with
+      | `Assoc list -> 
+        (match List.find list ~f:(fun (key, _) -> String.equal key name) with
+        | Some (_, v) -> Some v
+        | _ -> None)
+      | _ -> None in
+    try
+     let symbol_list = let open Yojson in
+      match R2.with_command_j "isj" file with
+      | `List list -> Some list
+      | _ -> None in
+     let strip str = let open String in
+      match chop_prefix str ~prefix:"sym.imp." with
+      | Some str -> str
+      | None -> str in
+     let open Yojson in
+     let to_string json = 
+      match json with
+      | `String str -> Some str
+      | _ -> None in
+     let to_int json = 
+      match json with
+      | `Int i -> Z.of_int i |> Some
+      | `Intlit s -> Z.of_string s |> Some
+      | _ -> None in
+     Option.(symbol_list >>| List.fold ~init:() ~f:(fun () symbol -> 
+            both
+            (extract "name" symbol >>= to_string >>| strip) 
+            (extract "vaddr" symbol >>= to_int) 
+            |> value_map ~default:() ~f:(fun (name, addr) -> accept name addr)
+     ) |> value ~default:());
+     if Hashtbl.length funcs = 0
+     then warning "failed to obtain symbols";
+     let symbolizer = Symbolizer.create @@ fun addr ->
+     Hashtbl.find funcs @@
+     Bitvec.to_bigint (Word.to_bitvec addr) in
+     Symbolizer.provide agent symbolizer;
+     provide_roots funcs
+    with _ -> warning "failed to use radare2";()
```

Comment:
Style note: limit the scope `try/catch` statements to the scope of the expressions that you know can fail and never hush bugs and other programmer errors with `try/catch` statements. Also, deal with errors as soon as possible, to not let them propagate to far away from the point of origin. 

In BAP and Core we don't write functions that casually throw exceptions and rely on explicit algebraic data types to denote functions that are not partial. We keep exceptions for programmers errors and never hush them, so that if we introduce a bug it is immediately caught by our CI system with a readable backtrace. 

In this case the only function that can fail with a runtime error is `R2.with_command` so we need to handle it correctly, we can use the ability to match on exceptions of the `match` construct, e.g.,

```ocaml
match R2.with_command_j "isj" file with
  | `List list -> list
  | s -> warning "unexpected output: %a" Yojson.pp s
  | exception Invalid_argument msg -> warning "failed to get the symbols: %s" msg
```

Note, that we removed an unnecessary wrapping of the list into the option type, which will simplify the logic downstream, as well as we let all exceptions other than `Invalid_argument` to propagate further so that we do not hash unexpected errors and accumulate problems. We're catching `Invalid_argument` as, although it is undocumented, it is the exception that `R2` is using to indicate runtime errors. A yet another argument for using algebraic data types instead of exceptions to report erroneous conditions.


Also, `f x; ()` is the same as `f x`.